src/data_structure/graph.rs

```diff
@@ -0,0 +1,221 @@
+use std::collections::HashMap;
+use std::collections::HashSet;
+use std::marker::PhantomData;
+
+/// 'Graph<N, E, Ty>' is a data structure for graphs with optionally
+/// weighted nodes and edges of arbitrary type. This implementation
+/// is inteded for educational purposes only. 
+/// See [petgraph](https://docs.rs/crate/petgraph/) for a full featured
+/// implementation.
+pub struct Graph<N, E, Ty = Directed> {
+    nodes: HashMap<usize, Node<N>>,
+    edges: HashMap<usize, Edge<E>>,
+    ty: PhantomData<Ty>,
+}
+
+impl<N, E, Ty> Graph<N, E, Ty>
+where
+    Ty: EdgeType,
+{
+    fn new() -> Self {
+        Graph {
+            nodes: HashMap::new(),
+            edges: HashMap::new(),
+            ty: PhantomData,
+        }
+    }
+
+    /// Returns true iff graph is directed.
+    fn is_directed(&self) -> bool {
+        Ty::is_directed()
+    }
+
+    /// Add node to graph and return its assigned index.
+    fn add_node(&mut self, weight: N) -> usize {
+        let node = Node { weight: weight };
+        let mut index = 0usize;
+        while self.nodes.contains_key(&index) {
+            index += 1;
+        }
+        self.nodes.insert(index, node);
+        index
+    }
+
+    /// Given two indices head and tail, attempt to add an edge whose
+    /// starting node has index head and whose ending node has index
+    /// tail to the graph. If this succeeds, return index assigned to
+    /// that edge. Otherwise return None.
+    fn add_edge(&mut self, weight: E, head: usize, tail: usize) -> Option<usize> {
+        if !self.nodes.contains_key(&head) || !self.nodes.contains_key(&head) {
+            None
+        } else {
+            let mut index = 0usize;
+            while self.edges.contains_key(&index) {
+                index += 1;
+            }
+            self.edges.insert(
+                index,
+                Edge {
+                    weight: weight,
+                    head: head,
+                    tail: tail,
+                },
+            );
+            Some(index)
+        }
+    }
+
+    /// Return first index of an edge of the form head --> tail.  This
+    /// is not necessarily the unique index since we explicitly allow
+    /// multiple edges between two nodes.
+    fn find_edge(&self, head: usize, tail: usize) -> Option<usize> {
+        for (index, edge) in self.edges.iter() {
+            // The latter part of the following disjunction is needed
+            // to identify [head,tail] with [tail,head] in undirected
+            // graphs.
+            if [edge.head, edge.tail] == [head, tail]
+                || (self.is_directed() && [edge.head, edge.tail] == [tail, head])
```

Comment:
This code look duplicate as in `find_edges`, can we reuse some of this?

Reply:
You're right. I'll get rid of that duplicate code.